File: SaigeIncPy2Cs/OptionsExperiment.cs
// 
// 
// refs
// # https://www.quantconnect.com/tutorials/strategy-library/volatility-risk-premium-effect
// # https://www.quantconnect.com/forum/discussion/2894/the-options-trading-strategy-based-on-macd-indicator/p1
// # https://www.quantconnect.com/tutorials/tutorial-series/applied-options
// # https://www.quantconnect.com/forum/discussion/5709/optionchain-is-empty/p1
// 

using timedelta = datetime.timedelta;

using np = numpy;

using pd = pandas;

using stats = scipy.stats;

using Accord.Statistics;
using System.Linq;
using System.Collections.Generic;
using QuantConnect.Algorithm.Framework.Alphas;
using QuantConnect.Algorithm.Framework.Execution;
using QuantConnect.Algorithm.Framework.Portfolio;
using QuantConnect.Algorithm.Framework.Risk;
using QuantConnect.Algorithm.Framework.Selection;
using QuantConnect.Interfaces;
using QuantConnect.Data.UniverseSelection;
using QuantConnect.Algorithm;
using QuantConnect;
using QuantConnect.Securities;
using QuantConnect.Data;
using QuantConnect.Orders;
using System;

public class OptionRouletteAlgorithm
        : QCAlgorithm {
        
        public List<object> atm_call;
        
        public List<object> atm_put;
        
        public List<object> otm_call;
        
        public List<object> otm_put;
        
        public None slice;
        
        public Symbol symbol;
        
        public override void Initialize() {
            this.SetStartDate(2017, 1, 15);
            this.SetEndDate(2017, 2, 15);
            //self.SetStartDate(2015, 1, 1)
            //self.SetEndDate(datetime.now().date() - timedelta(1))
            this.SetCash(100000);
            var equity = this.AddEquity("SPY", Resolution.Minute);
            var option = this.AddOption("SPY", Resolution.Minute);
            this.symbol = equity.Symbol;
            option.SetFilter(this.UniverseFunc);
            this.SetBenchmark(equity.Symbol);
            this.slice = null;
            // Define the Schedules
            this.Schedule.On(this.DateRules.WeekStart(this.symbol), this.TimeRules.AfterMarketOpen(this.symbol, 5), this.MyLiquidate);
            // Define the Schedules
            this.Schedule.On(this.DateRules.WeekStart(this.symbol), this.TimeRules.AfterMarketOpen(this.symbol, 10), this.MyTrade);
        }
        
        public override void OnData(Slice slice) {
            this.slice = slice;
            if (slice.OptionChains.Count > 0) {
            }
        }
        
        public override void OnAssignmentOrderEvent(OrderEvent assignmentEvent) {
            this.Log(assignmentEvent.ToString());
            this.MyLiquidate();
        }
        
        public override void OnOrderEvent(OrderEvent orderEvent) {
            this.Log(orderEvent.ToString());
        }
        
        public OptionFilterUniverse UniverseFunc(OptionFilterUniverse universe) {
            var price = this.Securities[this.symbol].Price;
            return universe.IncludeWeeklys().Strikes(-50, 50).Expiration(new timedelta(30), new timedelta(50));
            // TODO: read above api.
        }
        
        public void MyLiquidate() {
            foreach (var x in this.Portfolio) {
                if (x.Value.Invested) {
                    this.Liquidate(x.Key);
                }
            }
            // redundant?
            if (this.Portfolio[this.symbol].Invested) {
                this.Liquidate(this.symbol);
            }
            this.Log("MyLiquidate");
        }
        
        public void MyTrade() {
            object otm_put_strike;
            object otm_call_strike;
            object mylist;
            var slice = this.slice;
            if (slice == null) {
                return;
            }
            this.Log("MyTrade {} {}".Format(this.Portfolio.Invested, slice.OptionChains.Count));
            if (slice.OptionChains.Count == 0) {
                return;
            }
            foreach (var i in slice.OptionChains) {
                var chains = i.Value;
                if (!this.Portfolio.Invested) {
                    this.Log("trading!");
                    // divide option chains into call and put options 
                    var calls = chains.Where(x => x.Right == OptionRight.Call).ToList().ToList();
                    var puts = chains.Where(x => x.Right == OptionRight.Put).ToList().ToList();
                    // if lists are empty return
                    if (!calls || !puts) {
                        return;
                    }
                    var underlying_price = this.Securities[this.symbol].Price;
                    var expiries = (from i in puts
                        select i.Expiry).ToList();
                    // determine expiration date nearly one month
                    var expiry = Math.Min(expiries, key: x => Math.Abs((x.date() - this.Time.Date).days - 40));
                    var strikes = (from i in puts
                        select i.Strike).ToList();
                    // determine at-the-money strike
                    var strike = Math.Min(strikes, key: x => Math.Abs(x - underlying_price));
                    // compute probability
                    var hist = this.History(new List<Symbol> {
                        this.symbol
                    }, 252 * 5, Resolution.Daily);
                //var prct_changes = hist.loc[this.symbol]["close"].pct_change(40);
                Slice sliceHist = hist.FirstOrDefault();
                var prct_changes = hist.FirstOrDefault().
                // 68% = 1sd, 90% = 2sd.
                var _tup_1 = np.nanpercentile(prct_changes, new List<int> {
                        5,
                        32,
                        68,
                        95
                    });
                    var m2sd = _tup_1.Item1;
                    var m1sd = _tup_1.Item2;
                    var p1sd = _tup_1.Item3;
                    var p2sd = _tup_1.Item4;
                    // roulette logic
                    var optionStyle = np.random.choice(new List<string> {
                        "short_strangle",
                        "short_iron_condor",
                        "long_strangle",
                        "synthetic_long"
                    }, 1)[0];
                    var num = np.random.choice(new List<int> {
                        2,
                        5,
                        10
                    }, 1)[0];
                    // long volatility strategies ********************************
                    // why would you ever?
                    if (optionStyle == "synthetic_long") {
                        this.atm_put = (from i in puts
                            where i.Expiry == expiry && i.Strike == strike
                            select i).ToList();
                        this.atm_call = (from i in calls
                            where i.Expiry == expiry && i.Strike == strike
                            select i).ToList();
                        if (this.atm_put && this.atm_call) {
                            mylist = new List<List<object>> {
                                this.atm_put[0],
                                this.atm_call[0]
                            };
                            this.Log("{}".format((from x in mylist
                                select stats.percentileofscore(prct_changes, (x.Strike - underlying_price) / underlying_price)).ToList()));
                            this.Sell(this.atm_put[0].Symbol, num);
                            this.Buy(this.atm_call[0].Symbol, num);
                        }
                    }
                    if (optionStyle == "long_strangle") {
                        this.atm_put = (from i in puts
                            where i.Expiry == expiry && i.Strike == strike
                            select i).ToList();
                        this.atm_call = (from i in calls
                            where i.Expiry == expiry && i.Strike == strike
                            select i).ToList();
                        if (this.atm_put && this.atm_call) {
                            mylist = new List<List<object>> {
                                this.atm_put[0],
                                this.atm_call[0]
                            };
                            this.Log("{}".format((from x in mylist
                                select stats.percentileofscore(prct_changes, (x.Strike - underlying_price) / underlying_price)).ToList()));
                            this.Buy(this.atm_put[0].Symbol, num);
                            this.Buy(this.atm_call[0].Symbol, num);
                        }
                    }
                    // short volatility strategies ********************************
                    if (optionStyle == "short_iron_condor") {
                        otm_call_strike = min(strikes, key: x => Math.Abs(x - underlying_price + p2sd * underlying_price));
                        var atm_call_strike = min(strikes, key: x => Math.Abs(x - underlying_price + p1sd * underlying_price));
                        var atm_put_strike = min(strikes, key: x => Math.Abs(x - underlying_price + m1sd * underlying_price));
                        otm_put_strike = min(strikes, key: x => Math.Abs(x - underlying_price + m2sd * underlying_price));
                        this.otm_call = (from i in calls
                            where i.Expiry == expiry && i.Strike == otm_call_strike
                            select i).ToList();
                        this.atm_call = (from i in calls
                            where i.Expiry == expiry && i.Strike == atm_call_strike
                            select i).ToList();
                        this.atm_put = (from i in puts
                            where i.Expiry == expiry && i.Strike == atm_put_strike
                            select i).ToList();
                        this.otm_put = (from i in puts
                            where i.Expiry == expiry && i.Strike == otm_put_strike
                            select i).ToList();
                        if (this.atm_call && this.atm_put && this.otm_put && this.otm_call) {
                            mylist = new List<List<object>> {
                                this.otm_put[0],
                                this.atm_call[0],
                                this.atm_put[0],
                                this.otm_call[0]
                            };
                            this.Log("{}".format((from x in mylist
                                select stats.percentileofscore(prct_changes, (x.Strike - underlying_price) / underlying_price)).ToList()));
                            // TODO: log net profit and potential max loss.
                            // buy otm
                            this.Buy(this.otm_call[0].Symbol, num);
                            this.Buy(this.otm_put[0].Symbol, num);
                            // sell near atm
                            this.Sell(this.atm_call[0].Symbol, num);
                            this.Sell(this.atm_put[0].Symbol, num);
                        }
                    }
                    if (optionStyle == "short_strangle") {
                        otm_call_strike = min(strikes, key: x => Math.Abs(x - underlying_price + p2sd * underlying_price));
                        otm_put_strike = min(strikes, key: x => Math.Abs(x - underlying_price + m2sd * underlying_price));
                        this.otm_put = (from i in puts
                            where i.Expiry == expiry && i.Strike == otm_put_strike
                            select i).ToList();
                        this.otm_call = (from i in calls
                            where i.Expiry == expiry && i.Strike == otm_call_strike
                            select i).ToList();
                        if (this.otm_put && this.otm_call) {
                            mylist = new List<List<object>> {
                                this.otm_put[0],
                                this.otm_call[0]
                            };
                            this.Log("{}".format((from x in mylist
                                select stats.percentileofscore(prct_changes, (x.Strike - underlying_price) / underlying_price)).ToList()));
                            this.Sell(this.otm_put[0].Symbol, num);
                            this.Sell(this.otm_call[0].Symbol, num);
                        }
                    }
                }
            }
        }
}

File: SaigeIncPythonExamples/OptionsExperiment.py
"""

refs
# https://www.quantconnect.com/tutorials/strategy-library/volatility-risk-premium-effect
# https://www.quantconnect.com/forum/discussion/2894/the-options-trading-strategy-based-on-macd-indicator/p1
# https://www.quantconnect.com/tutorials/tutorial-series/applied-options
# https://www.quantconnect.com/forum/discussion/5709/optionchain-is-empty/p1
"""

from datetime import timedelta
import numpy as np
import pandas as pd
from scipy import stats
np.random.seed(2020) # comment to make it a real roulette

class OptionRouletteAlgorithm(QCAlgorithm):

    def Initialize(self):
        
        self.SetStartDate(2017, 1, 15)
        self.SetEndDate(2017,2, 15)
        #self.SetStartDate(2015, 1, 1)
        #self.SetEndDate(datetime.now().date() - timedelta(1))
        
        self.SetCash(100000)
        equity = self.AddEquity("SPY", Resolution.Minute)
        option = self.AddOption("SPY", Resolution.Minute)
        self.symbol = equity.Symbol
        option.SetFilter(self.UniverseFunc)
        self.SetBenchmark(equity.Symbol)
        self.slice = None

        # Define the Schedules
        self.Schedule.On(
            self.DateRules.WeekStart(self.symbol),
            self.TimeRules.AfterMarketOpen(self.symbol, 5),
            Action(self.MyLiquidate)
        )
        
        # Define the Schedules
        self.Schedule.On(
            self.DateRules.WeekStart(self.symbol),
            self.TimeRules.AfterMarketOpen(self.symbol, 10),
            Action(self.MyTrade)
        )
        
    def OnData(self,slice):
        self.slice = slice
        if slice.OptionChains.Count > 0:
            pass
            
    def OnAssignmentOrderEvent(self, assignmentEvent):
        self.Log(str(assignmentEvent))
        self.MyLiquidate()
        
    def OnOrderEvent(self, orderEvent):
        self.Log(str(orderEvent)) 

    def UniverseFunc(self, universe):
        price = self.Securities[self.symbol].Price
        return universe.IncludeWeeklys()\
                    .Strikes(-50,50)\
                    .Expiration(timedelta(30), timedelta(50))
        # TODO: read above api.
                    
    def MyLiquidate(self):
        for x in self.Portfolio:
            if x.Value.Invested:
                self.Liquidate(x.Key)
        # redundant?
        if self.Portfolio[self.symbol].Invested:
            self.Liquidate(self.symbol)
                
        self.Log("MyLiquidate")
    
    
    def MyTrade(self):
        slice = self.slice
        
        if slice is None:
            return
        
        self.Log("MyTrade {} {}".format(self.Portfolio.Invested,slice.OptionChains.Count))
        if slice.OptionChains.Count == 0:
            return
        for i in slice.OptionChains:
            chains = i.Value
            
            if not self.Portfolio.Invested:
                self.Log("trading!")
                # divide option chains into call and put options 
                calls = list(filter(lambda x: x.Right == OptionRight.Call, chains))
                puts = list(filter(lambda x: x.Right == OptionRight.Put, chains))
                
                # if lists are empty return
                if not calls or not puts: return
                
                underlying_price = self.Securities[self.symbol].Price
                expiries = [i.Expiry for i in puts]
                
                # determine expiration date nearly one month
                expiry = min(expiries, key=lambda x: abs((x.date()-self.Time.date()).days-40))
                strikes = [i.Strike for i in puts]
                
                # determine at-the-money strike
                strike = min(strikes, key=lambda x: abs(x-underlying_price))
                
                # compute probability
                hist = self.History([self.symbol], 252*5, Resolution.Daily)
                prct_changes = hist.loc[self.symbol]['close'].pct_change(40)
                # 68% = 1sd, 90% = 2sd.
                m2sd,m1sd,p1sd,p2sd = np.nanpercentile(prct_changes,[5,32,68,95])
                
                # roulette logic
                optionStyle = np.random.choice(['short_strangle','short_iron_condor','long_strangle','synthetic_long'],1)[0]
                num = np.random.choice([2,5,10],1)[0]
                
                # long volatility strategies ********************************
                
                # why would you ever?
                if optionStyle == 'synthetic_long':
                    self.atm_put = [i for i in puts if i.Expiry == expiry and i.Strike == strike]
                    self.atm_call = [i for i in calls if i.Expiry == expiry and i.Strike == strike]
                    
                    if self.atm_put and self.atm_call:
                        mylist = [self.atm_put[0],self.atm_call[0]]
                        self.Log('{}'.format([stats.percentileofscore(prct_changes,(x.Strike-underlying_price)/underlying_price) for x in mylist]))
                        
                        self.Sell(self.atm_put[0].Symbol, num)
                        self.Buy(self.atm_call[0].Symbol, num)
                        
                if optionStyle == 'long_strangle':
                    
                    self.atm_put = [i for i in puts if i.Expiry == expiry and i.Strike == strike]
                    self.atm_call = [i for i in calls if i.Expiry == expiry and i.Strike == strike]
                    
                    if self.atm_put and self.atm_call:
                        mylist = [self.atm_put[0],self.atm_call[0]]
                        self.Log('{}'.format([stats.percentileofscore(prct_changes,(x.Strike-underlying_price)/underlying_price) for x in mylist]))
                        
                        self.Buy(self.atm_put[0].Symbol, num)
                        self.Buy(self.atm_call[0].Symbol, num)
                
                # short volatility strategies ********************************
                
                if optionStyle == 'short_iron_condor':
                    
                    otm_call_strike = min(strikes, key = lambda x:abs(x-underlying_price+p2sd*underlying_price))
                    atm_call_strike = min(strikes, key = lambda x:abs(x-underlying_price+p1sd*underlying_price)) # more like near atm
                    atm_put_strike = min(strikes, key = lambda x:abs(x-underlying_price+m1sd*underlying_price))
                    otm_put_strike = min(strikes, key = lambda x:abs(x-underlying_price+m2sd*underlying_price))

                    self.otm_call = [i for i in calls if i.Expiry == expiry and i.Strike == otm_call_strike]
                    self.atm_call = [i for i in calls if i.Expiry == expiry and i.Strike == atm_call_strike]
                    self.atm_put = [i for i in puts if i.Expiry == expiry and i.Strike == atm_put_strike]
                    self.otm_put = [i for i in puts if i.Expiry == expiry and i.Strike == otm_put_strike]
                    
                    if self.atm_call and self.atm_put and self.otm_put and self.otm_call:
    
                        mylist = [self.otm_put[0],self.atm_call[0],self.atm_put[0],self.otm_call[0]]
                        self.Log('{}'.format([stats.percentileofscore(prct_changes,(x.Strike-underlying_price)/underlying_price) for x in mylist]))
                        # TODO: log net profit and potential max loss.
                        
                        # buy otm
                        self.Buy(self.otm_call[0].Symbol, num)
                        self.Buy(self.otm_put[0].Symbol, num)
                        # sell near atm
                        self.Sell(self.atm_call[0].Symbol, num)
                        self.Sell(self.atm_put[0].Symbol, num)
                
                
                if optionStyle == 'short_strangle':
                    
                    otm_call_strike = min(strikes, key = lambda x:abs(x-underlying_price+p2sd*underlying_price))
                    otm_put_strike = min(strikes, key = lambda x:abs(x-underlying_price+m2sd*underlying_price))
                    
                    self.otm_put = [i for i in puts if i.Expiry == expiry and i.Strike == otm_put_strike]
                    self.otm_call = [i for i in calls if i.Expiry == expiry and i.Strike == otm_call_strike]
                    if self.otm_put and self.otm_call:
                        mylist = [self.otm_put[0],self.otm_call[0]]
                        self.Log('{}'.format([stats.percentileofscore(prct_changes,(x.Strike-underlying_price)/underlying_price) for x in mylist]))
                        
                        self.Sell(self.otm_put[0].Symbol, num)
                        self.Sell(self.otm_call[0].Symbol, num)
